src/state/opamEnv.ml

```diff
@@ -435,10 +440,52 @@ let string_of_update st shell updates =
       Printf.sprintf "%sset -gx %s %s;\n"
         (make_comment comment) k v
   in
-  let export = match shell with
+  match shell with
     | `zsh | `sh  -> sh
     | `fish -> fish
-    | `csh -> csh in
+    | `csh -> csh
+
+let source root ~shell f =
+  let file f = OpamFilename.to_string (OpamPath.init root // f) in
+  match shell with
+  | `csh ->
+    Printf.sprintf "source %s >& /dev/null || true\n" (file f)
+  | `fish ->
+    Printf.sprintf "source %s > /dev/null 2> /dev/null; or true\n" (file f)
+  | _ ->
+    Printf.sprintf "test -r %s && . %s > /dev/null 2> /dev/null || true\n"
+      (file f) (file f)
+
+let get_shell_content root ~shell ~eval_env fvar fcomplete =
+  let ielse else_opt = match else_opt with  None -> "" | Some e -> Printf.sprintf "else\n  %s" e in
+  let if_csh t e  = Printf.sprintf "if ( $?prompt ) then\n  %s%sendif\n" t @@ ielse e in
+  let if_fish t e = Printf.sprintf "if isatty\n  %s%send\n" t @@ ielse e in
+  let if_sh t e   = Printf.sprintf "if [ -t 0 ]; then\n  %s%sfi\n" t @@ ielse e in
+  let if_st, opam_env = match shell with
+    | `csh -> if_csh, opam_env_csh
+    | `fish -> if_fish, opam_env_fish
+    | `zsh -> if_sh, opam_env_zsh
+    | `sh -> if_sh, opam_env_sh
+  in
+  let variables = source root ~shell fvar in
+  let complete = OpamStd.Option.map (source root ~shell) fcomplete in
+  if eval_env then
+    let thenn =
+      match complete with
+      | None -> opam_env
+      | Some cplt -> Printf.sprintf "%s\n  %s" opam_env cplt
+    in
+    if_st thenn (Some variables)
+  else
+    Printf.sprintf "%s\n%s"
+      (export_in_shell shell ("OPAM_NOEVAL_ENV", string_of_bool @@ not eval_env,
```

Comment:
Should be `OPAMNOEVALENV`. Ugly, I agree, but consistent with all other env variables opam reads.